hawkbit-extension-artifact-repository-s3/src/main/java/org/eclipse/hawkbit/artifact/repository/S3Artifact.java

```diff
@@ -39,7 +41,21 @@
 
     @Override
     public InputStream getFileInputStream() {
-        return amazonS3.getObject(s3Properties.getBucketName(), key).getObjectContent();
+        final S3Object s3Object = amazonS3.getObject(s3Properties.getBucketName(), key);
+        if (s3Object == null) {
+            return null;
+        }
+        
+        try {
+            return s3Object.getObjectContent();
+        } finally {
+            // Make sure s3Object is closed to avoid memory leak
+            try {
```

Comment:
Unfortunately this doesn't work, you cannot do a finally block after a return which closes the stream. Because the finally block is executed before the return statement then and the called will retrieve a closed inputstream then and cannot read from it anymore. You can easily verify it by a small test like this:
```
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

import org.junit.Test;

public class StreamTest {

    @Test
    public void streamClosedStream() throws IOException {
        final byte[] buffer = new byte[1];
        getStream().read(buffer);
    }

    private InputStream getStream() throws IOException {
        final File tmpFile = File.createTempFile("streamtest", "");
        final InputStream inputStream = new FileInputStream(tmpFile);

        try {
            return inputStream;
        } finally {
            try {
                inputStream.close();
            } catch (final IOException e) {
                e.printStackTrace();
            }
        }
    }
}
```

That's the reason the `public abstract InputStream getFileInputStream()` says in the javadoc `Caller has to take care
   of closing the stream. Repeatable calls open a new {@link InputStream}.` which is also done in the hawkBit code.

Reply:
You are correct. I had dealt with some single-shot `InputStream`s not too long ago and inadvertently applied that logic here. The streams should be managed by the caller. I'll revise this change to close the S3 objects when the repository object is deleted.